torch/csrc/jit/init.cpp

```diff
@@ -379,16 +376,95 @@ void initJITBindings(PyObject* module) {
     });
   });
 
-  // NOLINTNEXTLINE(bugprone-unused-raii)
-  py::class_<detail::Future>(m, "Future");
 
-  m.def("fork", [](script::Module& sm, py::args args) {
-    // TODO: this is a fake stub
-    return detail::Future();
+  struct PythonFutureWrapper {
+    PythonFutureWrapper(c10::intrusive_ptr<c10::ivalue::Future> fut) : fut(std::move(fut)) {}
+
+    c10::intrusive_ptr<c10::ivalue::Future> fut;
+  };
+
+
+  py::class_<PythonFutureWrapper>(m, "Future");
+
+  m.def("fork", [](py::args args) {
+    AT_ASSERT(args.size() >= 1);
+
+    py::function f = py::cast<py::function>(args[0]);
+    py::tuple args_tup(args.size() - 1);
+
+    for (size_t i = 1; i < args.size(); ++i) {
+      args_tup[i - 1] = args[i];
+    }
+
+    if (jit::tracer::isTracing()) {
+      auto graph = jit::tracer::getTracingState()->graph;
+      auto fork_node = graph->insertNode(graph->create(prim::fork, 1));
+      auto body_block = fork_node->addBlock();
+
+      Value *node_output;
+      py::object py_func_output;
+      auto retval = c10::make_intrusive<c10::ivalue::Future>();
+      // Insert new trace ops into the fork op's sub-block
+      WithInsertPoint guard(body_block);
+      IValue output_ivalue;
+      {
+        tracer::WithTracingEnvStack env_guard;
+
+        // Run the user-supplied function
+        py_func_output = f(*args_tup);
+
+        // Convert the output of the user-supplied funciton to IValue. The type
+        // information of this IValue is used both to record the correct type in
+        // the trace.
+        output_ivalue = toIValue(py_func_output);
+        Value *out_val = jit::tracer::getNestedValueTrace(output_ivalue);
+        body_block->registerOutput(out_val);
+        node_output = fork_node->output()->setType(FutureType::create(out_val->type()));
+
+        // Lambda lift
```

Comment:
This is a repeat from compiler.cpp, can we make a common function, `lambdaLiftFork(Node* fork_node)`